controllers/datadogmonitor/controller.go

```diff
@@ -0,0 +1,318 @@
+	maxTriggeredStateGroups = 10
+)
+
+// Reconciler reconciles a DatadogMonitor object
+type Reconciler struct {
+	client        client.Client
+	datadogClient *datadogapiclientv1.APIClient
+	datadogAuth   context.Context
+	versionInfo   *version.Info
+	log           logr.Logger
+	scheme        *runtime.Scheme
+	recorder      record.EventRecorder
+}
+
+// NewReconciler returns a new Reconciler object
+func NewReconciler(client client.Client, ddClient datadogclient.DatadogClient, versionInfo *version.Info, scheme *runtime.Scheme, log logr.Logger, recorder record.EventRecorder) (*Reconciler, error) {
+	return &Reconciler{
+		client:        client,
+		datadogClient: ddClient.Client,
+		datadogAuth:   ddClient.Auth,
+		versionInfo:   versionInfo,
+		scheme:        scheme,
+		log:           log,
+		recorder:      recorder,
+	}, nil
+}
+
+// Reconcile is similar to reconciler.Reconcile interface, but taking a context
+func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	return r.internalReconcile(ctx, request)
+}
+
+// Reconcile loop for DatadogMonitor
+func (r *Reconciler) internalReconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	logger := r.log.WithValues("datadogmonitor", req.NamespacedName)
+	logger.Info("Reconciling DatadogMonitor")
+	now := metav1.NewTime(time.Now())
+
+	// Get instance
+	instance := &datadoghqv1alpha1.DatadogMonitor{}
+	var result ctrl.Result
+	err := r.client.Get(ctx, req.NamespacedName, instance)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return result, nil
+		}
+		// Error reading the object - requeue the request
+		return ctrl.Result{RequeueAfter: defaultErrRequeuePeriod}, err
+	}
+
+	newStatus := instance.Status.DeepCopy()
+
+	if result, err = r.handleFinalizer(logger, instance); err != nil || result.Requeue {
+		return r.updateStatusIfNeeded(logger, instance, now, newStatus, err, result)
+	}
+
+	// Validate the DatadogMonitor spec
+	if err = datadoghqv1alpha1.IsValidDatadogMonitor(&instance.Spec); err != nil {
+		logger.Error(err, "Invalid DatadogMonitor spec")
+		return r.updateStatusIfNeeded(logger, instance, now, newStatus, err, result)
+	}
+
+	instanceSpecHash, err := comparison.GenerateMD5ForSpec(&instance.Spec)
+	if err != nil {
+		logger.Error(err, "Error generating hash")
+		return r.updateStatusIfNeeded(logger, instance, now, newStatus, err, result)
+	}
+
+	statusSpecHash := instance.Status.CurrentHash
+
+	// Create or update monitor, or check monitor state. Fall through this block (without returning)
+	// if the result should be requeued with the default period
+	if instance.Status.ID == 0 {
+		// If the monitor ID is 0, then it doesn't exist yet in Datadog. Create the monitor (only metric alerts)
```

Comment:
This would be a great spot to log that we are about to create monitor?

Reply:
there's a log of when a monitor is created; think there should be one prior to creation as well?

Reply:
Not a big deal, but might be helpful in troubleshooting, but iirc there isn't Debug level, only Info?